app/ide-desktop/lib/dashboard/src/layouts/Editor.tsx

```diff
@@ -61,112 +56,48 @@ export default function Editor(props: EditorProps) {
-              history.replaceState(null, '', new URL('.', originalUrl))
-            }
-            try {
-              await appRunner.runApp(
-                {
-                  loader: {
-                    assetsUrl: `${assetsRoot}dynamic-assets`,
-                    wasmUrl: `${assetsRoot}pkg-opt.wasm`,
-                    jsUrl: `${assetsRoot}pkg${JS_EXTENSION[backendType]}`,
-                  },
-                  engine: engineConfig,
-                  startup: {
-                    project: project.packageName,
-                    displayedProjectName: project.name,
-                  },
-                  window: {
-                    topBarOffset: `${TOP_BAR_X_OFFSET_PX}`,
-                  },
-                },
-                accessToken,
-                {
-                  projectId: project.projectId,
-                  ignoreParamsRegex: new RegExp(`^${appUtils.SEARCH_PARAMS_PREFIX}(.+)$`),
-                }
-              )
-            } catch (error) {
-              toastAndLog('openEditorError', error)
-            }
-            if (backendType === backendModule.BackendType.remote) {
-              // Restore original URL so that initialization works correctly on refresh.
-              history.replaceState(null, '', originalUrl)
-            }
-          }
-          if (supportsLocalBackend) {
-            await runNewProject()
-          } else {
-            if (!initialized) {
-              await Promise.all([
-                load.loadStyle(`${assetsRoot}style.css`),
-                load
-                  .loadScript(`${assetsRoot}entrypoint.js.gz`)
-                  .catch(() => load.loadScript(`${assetsRoot}index.js.gz`)),
-              ])
-              setInitialized(true)
-            }
-            await runNewProject()
-          }
-        }
-      })()
-      return () => {
-        appRunner.stopApp()
-      }
-    } else {
-      return
+
+    return {
+      config: {
+        engine: {
+          rpcUrl: jsonAddress,
+          dataUrl: binaryAddress,
+          ydocUrl: ydocAddress,
+        },
+        startup: {
+          project: project.packageName,
+          displayedProjectName: project.name,
+        },
+        window: {
+          topBarOffset: `${TOP_BAR_X_OFFSET_PX}`,
+        },
+      },
+      projectId,
+      hidden,
+      ignoreParamsRegex: new RegExp(`^${appUtils.SEARCH_PARAMS_PREFIX}(.+)$`),
+      logEvent,
     }
-  }, [projectStartupInfo, toastAndLog, /* should never change */ appRunner])
+  }, [projectStartupInfo, toastAndLog, hidden])
 
-  return <></>
+  if (projectStartupInfo == null || appRunner == null || appProps == null) {
+    return <></>
+  }
+  const AppRunner = appRunner
+  // Currently the GUI component needs to be fully rerendered whenever the project is changed. Once
+  // this is no longer necessary, the `key` could be removed.
+  return <AppRunner {...appProps} key={appProps.projectId} />
```

Comment:
nit: i think our preference in dashboard is key as first prop (since it's not a regular prop, and it makes it more obvious when reading a .map from left to right)